### PR TITLE
drivers: Add interrupt controller API

### PR DESCRIPTION
**Pull request by bjarki-andreasen**
This PR has now been updated to implement the solution described in this issue https://github.com/zephyrproject-rtos/zephyr/issues/67583

Size comparisons:
```
                        IRQ             SHARED          DYNAMIC         LVL        LTO
IRQ FLASH           38244 B            39652 B          39648 B     39776 B    34436 B
IRQ RAM              5912 B             5912 B           7848 B      7848 B     7840 B
SYS IRQ FLASH       42276 B            42276 B          42676 B     42676 B    36776 B
SYS IRQ RAM          5912 B             5912 B           6104 B      6104 B     5912 B
```
(note that no dynamic irq is registered, so LTO optimizes out the dynamic irq list)

Some example output of the generated code :)
Generated handlers:
```
static inline void __sys_intl_handler_8_5(void)
{
	/*
	 * Interrupt generating device: /soc/peripheral@50000000/clock@5000
	 */
	if (__sys_irq_handler_8_5_75()) {
		return;
	}

	/*
	 * Interrupt generating device: /soc/peripheral@50000000/power@5000
	 */
	if (__sys_irq_handler_8_5_117()) {
		return;
	}

	if (sys_irq_dynamic_handler(1)) {
		return;
	}

	sys_irq_log_spurious_intl(8, 5);

	sys_irq_spurious_handler();
}
```
Generated IRQ enumerations:
```
/*
 * Interrupt controller: /soc/interrupt-controller@e000e100
 * Interrupt line: 3
 * Interrupt generating device: /soc/spu@50003000
 */
#define SYS_DT_IRQN_8_3 0

/*
 * Interrupt controller: /soc/interrupt-controller@e000e100
 * Interrupt line: 5
 * Interrupt generating device: /soc/peripheral@50000000/clock@5000
 * Interrupt generating device: /soc/peripheral@50000000/power@5000
 */
#define SYS_DT_IRQN_8_5 1
```

A very cool solution I want to note as well is this macro:
```
INTC_DT_DEFINE_IRQ_VECTOR(node_id)
{
        return 1;
}
```
which wraps the existing ARCH_ISR_DIRECT_DECLARE(name) macro synthesizing the name of the function to match the weakly defined default vector in the vector table, allowing the existing infrastructure to be used, including ISR_DIRECT_PM()

## For arch review
This PR introduces two new requirements:
* We need to tell the kernel to generate interrupt handlers for devices which don't have status "okay" to:
  * Generate interrupt handlers for reserved devices, like RTCs reserved for sys tick or other SoC specific features
  * Generate interrupt handlers for devices, but not enable a driver for unit testing.
* Interrupt handlers need to signal whether they handled an IRQ or not in case of:
  * Shared IRQs to inform the kernel when we can stop walking IRQ handlers
  * Shared and normal IRQs, to allow the kernel to detect spurious interrupts

To generate interrupt handlers, but exclude device drivers, the status "reserved" is used to mean "supported by kernel, don't include a driver for it".

IRQ handlers will signal whether they handled the IRQ or not by returning 1 if handled, 0 if not handled.

## Analyzing code

**INTC_DT_DEFINE_IRQ_VECTOR with and without LTO**
```
mov     r0, sp
bic.w   r1, r0, #7
mov     sp, r1
push    {r0, lr}
ldr     r0, [pc, #16]
bl      0x2b00 <k_sem_give> /* first call in direct isr body (irq handler) */
```

**SYS_DT_DEFINE_IRQ_HANDLER without LTO**
```
mov     r0, sp
bic.w   r1, r0, #7
mov     sp, r1
push    {r0, lr}
bl      75c <__sys_irq_handler_7_20_107>
push    {r0, r1, r2, lr}
ldr     r3, [pc, #32]   ; (780 <__sys_irq_handler_7_20_107+0x24>)
ldr     r2, [pc, #32]   ; (784 <__sys_irq_handler_7_20_107+0x28>)
str     r3, [sp, #4]
ldr     r3, [pc, #32]   ; (788 <__sys_irq_handler_7_20_107+0x2c>)
ldr     r1, [pc, #36]   ; (78c <__sys_irq_handler_7_20_107+0x30>)
str     r3, [sp, #0]
movs    r0, #1
movs    r3, #34 ; 0x22
bl      158 <z_zassert.constprop.0>
```

**SYS_DT_DEFINE_IRQ_HANDLER with LTO**
```
mov     r0, sp
bic.w   r1, r0, #7
mov     sp, r1
push    {r0, lr}
ldr     r3, [pc, #40]   ; (3f14 <__intc_vector_7_20+0x34>)
sub     sp, #8
str     r3, [sp, #4]
ldr     r3, [pc, #36]   ; (3f18 <__intc_vector_7_20+0x38>)
ldr     r2, [pc, #40]   ; (3f1c <__intc_vector_7_20+0x3c>)
ldr     r1, [pc, #40]   ; (3f20 <__intc_vector_7_20+0x40>)
str     r3, [sp, #0]
movs    r0, #1
movs    r3, #34 ; 0x22
bl      581c <z_zassert.constprop.0>
```

**ISR_DIRECT_DECLARE without LTO**
```
mov     r0, sp
bic.w   r1, r0, #7
mov     sp, r1
push    {r0, lr}
ldr     r0, [pc, #16]   ; (1d0 <CONFIG_SRAM_SIZE+0x10>)
bl      4810 <z_impl_k_sem_give>
```

**ISR_CONNECT with and without LTO**
```
push    {r3, lr}
mrs     r3, IPSR
ldr     r2, [pc, #24]   ; (1500 <_isr_wrapper+0x20>)
subs    r3, #16
add.w   r1, r2, r3, lsl #3
ldr.w   r0, [r2, r3, lsl #3]
ldr     r1, [r1, #4]
blx     r1
```

### CODE REVIEW

**Comment by carlescufi** [comment]
Arch WG meeting:

- @bjarki-trackunit presents the proposal
- @carlescufi asks if the current `irq_enable()`-like APIs will be completely replaced by `intc_` ones. @bjarki-trackunit answers that yes, they will
- @gmarull asks if the compile-time checks will remain. @bjarki-trackunit says yes, with the help of the single-linked list
- @carlescufi mentions direct ISRs and the current interrupt model with the software ISR table
- @nashif mentions that the IRQ code needs to be available very early on (even pre-C)
- @gmarull asks about the Devicetree layout. One of the issues is about GPIO being both a GPIO controller and an interrupt controller. @bjarki-trackunit explains how he's solved it
-

File: drivers/intc/intc_arm_v8m_nvic.c [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+	INTC_DT_ARM_V8M_NVIC_DIRECT_IRQ_HANDLER_DEFINE(arm_v8m_nvic_default_handler##inst)	\
+	{											\
+		uint32_t irq_num = arm_v8m_nvic_get_active_irq_num();				\
+		struct intc_irq **irqs = arm_v8m_nvic_irqs##inst;				\
+		struct intc_irq *irq = irqs[irq_num];						\
+												\
+		if (irq != NULL) {								\
+			irq->handler(irq);							\
+		} else {									\
+			while (1) {								\
```
Comment (andyross) [inline drivers/intc/intc_arm_v8m_nvic.c:166]: Not understanding this at all.  Presumably you get here on a spurious interrupt?  We have an existing spurious interrupt handler API the arch layers all honor, as that's a pretty routine glitch during device bringup (e.g. "oops typo in the interrupt number", or "oops, the bootloader left the UART RX interrupt enabled" -- absolutely routine things).  Entering a silent infinite loop seems like exactly the wrong thing to do.

File: include/zephyr/drivers/intc.h [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+	uint16_t irq;
+	/** Interrupt request priority */
+	uint16_t priority;
+	/** Interrupt request flags */
+	uint32_t flags;
+};
+
+/** Interrupt request handle */
+struct intc_irq {
+	/** Interrupt request single-linked list node */
```
[the remark this answers is not in the document]
Comment (andyross) [reply include/zephyr/drivers/intc.h:274]: Actually, it doesn't go anywhere?  I can't find code that uses this.  Is this a forgotten attempt to implement shared interrupts?  As of right now it doesn't seem to work?

**Review by andyross — COMMENTED** [review]
So... I don't hate the API, it's clean and reasonable (though it's also wordy, requiring three steps to do what a single IRQ_CONNECT() does right now).

But this seems like only about 1/3 of the solution.  This is all about the user/driver/subsystem-facing API.  But the API we have for that is really... not so bad right now.  What's the story for why we want to merge this?  What can we do with it that we can't right now?

The problems with our interrupt layer (and it has problems!) are on the other side:

* There's no management of interrupt controller cascade by the framework.  Every platform needs to write its own glue to call child interrupt controller handlers out of their parents, even though the structure of that tree is known at build time by devicetree.  This PR seems targetted at only NVIC platforms without a secondary controller, meaning that it's really not seeing the mess we're currently living with.

* The interrupt ID/number/index/offset management (i.e. the mapping between a given device interrupt and the index into the _sw_isr_table[] array) is likewise ad-hoc and platform dependent, and is generally implemented with a bunch of crazy magic numbers stored in kconfig.  And again, a clean mapping could be trivially derived from the graph of controllers known to DTS.

Also, it's worth pointing out that this scheme is throwing out efficiency features we've had for a long time: Interrupts are 100% runtime now, you have to write and call C code to get an interrupt registered, spending bytes on code that has to execute and RAM on data structures that have to be mutated.  Some platforms even today are still RAM-constrained and would like their ISR tables in pre-built flash.  Maybe we don't care about apps like that anymore, it's an argument worth having.  But we should be clear about what we're walking away from.

tl;dr: I guess my feeling is that if we're going to rewrite interrupt handling, we should sharpen the machetes and do it all, for everyone.

**Comment by andyross** [comment]
Looks like the bot didn't assign reviewers, maybe because it's a draft?  For sure @dcpleung @peter-mitsis @carlocaione @npitre want to get in on this.  @tbursztyka should expand my note about MSI.   Lots of platform folks will have important input too, especially ones who've had to deal with cascaded controllers.

**Comment by bjarki-trackunit** [comment]
@andyross This API aims to solve points 1. and 2. while adding support for "secondary" interrupt controllers, like GPIO ports and other IO expanders which can route interrupts through IO pins.

1. Interrupt cascading
This API specifies every IRQ as an interrupt line of a specific interrupt controller, like a GPIO pin is a pin of a specific GPIO port. This reflects the devicetree (hardware) layout completely, as opposed to the current "absolute" ISR table, which tries to flatten the layout into a single list, which has spawned the need for encoding the IRQ number in levels.

2. IRQ encoding
This solution does not require IRQ encoding, as each IRQ number is relative to an the IRQs interrupt parent. Each parent manages its own list of IRQs in whicever way is most efficient. There is no absolute ISR table using this API, there can be as many tables as there are interrupt controllers, some interrupt controllers (like GPIOs) may simply store the few IRQs they manage in a single linked list to save memory as they do now.

3. Efficiency
This solution does remove the possibility of storing the SW IRQ table in ROM, but it can be quite a bit more efficient at storing IRQs, given that GPIOs may store their IRQs in single linked lists as they do now rather than an array of however many pins they have, for a chip with 4 ports of 32 pins, that'quite a bit.

For really ressource constrained devices, direct IRQs can be used, without the need for creating a SW IRQ table. Direct IRQs can be created for every interrupt controller at any level with this API, including a GPIO pin :)

File: include/zephyr/drivers/intc.h [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+
+/**
+ * @brief Get the name of an interrupt request handler
+ *
+ * @param ord Interrupt controller's devicetree ordinal
+ * @param irq Interrupt line number
+ *
+ * Expands to __intc_irq_handler_<ord>_<irq>
+ */
+#define INTC_IRQ_HANDLER_NAME(ord, irq) \
```
Comment (rakons) [inline include/zephyr/drivers/intc.h:429]: Is this meant to be used anywhere else than in the macro below? I am asking because it requires very strict values here and it seems little risky to reveal the macro as an part of the API.

Comment (bjarki-trackunit) [reply include/zephyr/drivers/intc.h:429]: Its meant to be used within device drivers to populate their interrupt vector tables, maybe it can be moved to a less user facing header like direct_irq_handler.h :)

File: include/zephyr/drivers/intc.h [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+/** Interrupt request handle forward declaration */
+struct intc_irq;
+
+/** Interrupt request handler */
+typedef void (*intc_irq_handler_t)(struct intc_irq *irq);
+
+/** @cond INTERNAL_HIDDEN */
+
+/** Interrupt request specification */
+struct intc_irq_spec {
```
[the remark this answers is not in the document]
Comment (bjarki-trackunit) [reply include/zephyr/drivers/intc.h:261]: The issue with the old-style number is that while being simple, it is lacking vital information, namely which interrupt controller the interrupt belongs to. The old style number is therefore only reflective of the devicetree, allowing `DT_IRQN` to return the correct IRQ number, if there is only one interrupt controller, at which point the instance can be omitted, and the API can just be static (as it is now).

An interrupt number is the interrupt line number of a specific interrupt controller. To clearly visualize the issue, consider how you would rewrite the GPIO API to only use an integer for the pin number, with a single API.

[the remark this answers is not in the document]
Comment (andyross) [reply include/zephyr/drivers/intc.h:261]: > it is lacking vital information

It's not.  Or rather it doesn't have to be.  The controller is extractable if you know how the table is packed, which right now requires interpreting a bunch of kconfig values storing sizes and offets, and it's a mess.  Which is why I'd like to see it reworked.

But that's not an indictment of using an integer.  We can totally continue to use integers to identify interrupts.  And (implemented correctly) integers have some excellent efficiency benefits which would be good to preserve.

(Also, just to repeat: this API requires a separate step to construct and extract that spec struct, vs. just using a by-value macro like DT_IRQN().  That right off the bat makes this a **worse** API, not a better one.  It's not a critical problem, but it's still a cost we're paying for exposing the spec abstraction to the user)

File: include/zephyr/drivers/intc.h [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+	/** Interrupt line's interrupt controller */
+	const struct device *intc;
+	/** Interrupt controller's interrupt line number  */
+	uint16_t irq;
+	/** Interrupt request priority */
+	uint16_t priority;
+	/** Interrupt request flags */
+	uint32_t flags;
```
[the remark this answers is not in the document]
Comment (bjarki-trackunit) [reply include/zephyr/drivers/intc.h:269]: the priority and irq number are integers, what will probably never exceed an 65365, but do exceed 255, where the flags is a bitfield, so it should be as large as we can provide :)

That said, I may update flags to be a size_t, but it does seem MSI fields can fit inside a uint32 https://i.stack.imgur.com/RwenJ.jpg

File: include/zephyr/drivers/intc.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (bjarki-trackunit) [reply include/zephyr/drivers/intc.h:274]: Its supposed to be used for shared IRQs, and for secondary interrupt controllers like GPIOs, which usually only store a few IRQs relative to how many pins they have.

File: include/zephyr/drivers/intc.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (bjarki-trackunit) [reply include/zephyr/drivers/intc.h:261]: Can you see the parallel I'm drawing between the GPIO API and the IRQ API? We have no problem pairing a GPIO controller to a pin and passing this to the GPIO API (in fact, its rather simple and elegant IMO), why treat IRQs differently?

File: include/zephyr/drivers/intc.h [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+
+/** Trigger interrupt request API template function */
+typedef int (*intc_api_trigger_irq)(const struct intc_irq_spec *spec);
+
+/** Release interrupt request API template function */
+typedef int (*intc_api_release_irq)(struct intc_irq *irq);
+
+/** Interrupt controller API structure */
+struct intc_driver_api {
+	intc_api_request_irq request_irq;
```
[the remark this answers is not in the document]
Comment (bjarki-trackunit) [reply include/zephyr/drivers/intc.h:302]: I think it looks good, it may be more lines, but every API in the struct does not have to be split into multiple lines :)

File: include/zephyr/drivers/intc.h [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+		.flags = INTC_DT_GET_IRQ_FLAGS_BY_NAME(node_id, name),		\
+		.priority = INTC_DT_GET_IRQ_PRIORITY_BY_NAME(node_id, name),	\
+	}
+
+/**
+ * @brief Get IRQ devicetree specifier by node identifier and name if IRQ exists
+ * @note Identical to INTC_DT_GET_IRQ_SPEC_BY_NAME() if IRQ exists
+ * @see INTC_DT_GET_IRQ_SPEC_BY_NAME()
+ */
+#define INTC_DT_GET_OPT_IRQ_SPEC_BY_NAME(node_id, name)				\
```
[the remark this answers is not in the document]
Comment (bjarki-trackunit) [reply include/zephyr/drivers/intc.h:187]: bus attached devices like sensors and RTCs are the rational for this API. Sensors often have 2-3 GPIO routed optional APIs, and there may be more than one instance of them, making `IS_ENABLED` and similar quite complex. See https://github.com/zephyrproject-rtos/zephyr/blob/30c660162f5fe70606bd122b61818ace47e6d908/drivers/sensor/lis2dh/lis2dh.c#L515-L526 for example (and note that this API is intended to replace the interrupt part of the GPIO API)

**Comment by rakons** [comment]
I am not sure if I get the implementation correctly. What I get from here, is that it is trying to solve the issue with multi level interrupts and how they are implemented now. I cannot quite get if does it replace the whole current irq implementation or only for the nested controllers?
If it is planned to be used as a general API for top level interrupts I can see 2 main issues here:
1. Breaking the API. This gives us not only a lot of work on the Zephyr level, but also all the drivers, here and in external modules, would have to be updated.
2. Lack of the possibility to put SWI in ROM - for some of our smaller chips I can see it as a quite challenging to fit everything we need in RAM when we include SWI table here. I know that now we have some samples where a lot of work is put into cutting the RAM or ROM usage.

And as we could live with the need of API change, lacking the possibility to use ROM for SWI table may be serious drawback for some bigger samples.

For us (but please note it would be extremely biased) - only pros I can see are:
- The script to generate SWI tables is no longer required. So one step that can make troubles less.
- It does not block LTO usage - but this one we can have with the same cost of switching from ROM to RAM by the usage of dynamic interrupts, without the cost of API break.

It does not seem to conflict with #66392. At least not in a code level. But at the API level it might be an issue. I suppose, that we could optimize how nested interrupt controllers are handled using (linker) similar approach like in mentioned PR, but still encoding the nested interrupt in interrupt number. But I am also not 100% sure if this is the main goal of this PR.

@bjarki-trackunit - do you see any problem with merging both of this PRs? On code level, I am not even thinking there would be any conflict reported by git. But from conceptual point of view they seem to be totally different beasts.

**Comment by bjarki-trackunit** [comment]
## Notes
### Fully implemented shared and dynamic interrupts
To properly support shared and dynamic interrupts, a return value has been introduced to the IRQ handlers. If an IRQ handler handled the IRQ, it returns 1, otherwise 0. An example interrupt line handler follows:
```
static inline void __sys_intl_handler_8_5(void)
{
	/*
	 * Interrupt generating device: /soc/peripheral@50000000/clock@5000
	 */
	if (__sys_irq_handler_8_5_75()) {
		return;
	}

	/*
	 * Interrupt generating device: /soc/peripheral@50000000/power@5000
	 */
	if (__sys_irq_handler_8_5_117()) {
		return;
	}

	if (sys_irq_dynamic_handler(1)) {
		return;
	}

	sys_irq_log_spurious_intl(8, 5);

	sys_irq_spurious_handler();
}
```
The static IRQ handlers are attempted first, if none of them where the originator of the interrupt, the list of dynamic interrupts is iterated. If no registered or linked IRQ handler could handle the IRQ, the kernel will invoke a spurious IRQ fault.

### Spurious IRQ logging added
Spurious interrupts now produce the following output:
```
[00:00:00.411,224] <err> os: >>> ZEPHYR SPURIOUS IRQ: INTC ord: 8, INT line: 21
[00:00:00.411,285] <err> os: >>> ZEPHYR FATAL ERROR 1: Unhandled interrupt on CPU 0
[00:00:00.411,315] <err> os: Current thread: 0x200006a8 (unknown)
[00:00:00.434,722] <err> os: Halting system
```
This error is "kernel level", meaning it is generic across all architectures, interrupt controllers etc. Simply look to the devicetree_generated.h to find the problematic interrupt controller by devicetree ordinal, them follow the irq number :)

File: drivers/intc/intc_arm_v8m_nvic/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) 2024 Bjarki Arge Andreasen
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library()
+
+zephyr_library_sources(arm_v8m_nvic.c)
+zephyr_linker_sources(ROM_START SORT_KEY 0x0vectors irq-vector-table.ld)
```
Comment (pdgendt) [inline drivers/intc/intc_arm_v8m_nvic/CMakeLists.txt:7]: Will this be an issue for the `CMAKE_LINKER_GENERATOR`?

Comment (bjarki-andreasen) [reply drivers/intc/intc_arm_v8m_nvic/CMakeLists.txt:7]: I think so, I will need to look into making it more generic, it is a bit of a hack currently :)

File: include/zephyr/drivers/intc.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (tbursztyka) [reply include/zephyr/drivers/intc.h]: Seems like this is fine for MSI.

Anyway, the implementation of MSI/MSI-X is currently very much arch specific, that is: x86 specific. So there won't be any collision with such generic intc API. 

It could change someday, since interrupt management implementation in x86 is messy to say the least (you always need to allocate an irq to get a vector for instance, that does not make much sense in case of MSI-X or multi-vector MSI).

**Comment by bjarki-andreasen** [comment]
Hope to continue this within a couple of weeks

**Comment by bjarki-andreasen** [comment]
## Update

* Fixed issue with LTO causing spurious interrupts (LTO optimized away entire vector table, fixed with __used attr)
* Cleaned up sys_irq test suites
* Rebased on latest main

**Comment by bjarki-andreasen** [comment]
## Update
* Added test suite for direct IRQs
* Cleaned up a bit

**Comment by bjarki-andreasen** [comment]
## Update

* Added `INTC_DT_FOREACH_INTL` and `INTC_DT_FOREACH_IMPL_INTL` macros for iterating over interrupt lines by interrupt controller. `INTC_DT_FOREACH_INTL` iterates over every interrupt line, whether any IRQ is implemented or not, while `INTC_DT_FOREACH_IMPL_INTL` only iterates over interrupt lines which are implemented (connected to node with status okay or reserved)
* Updated NVIC drivers to use the new foreach macros

**Comment by bjarki-andreasen** [comment]
## Update

* Quite a bit of cleanup
* Added macros for getting the number of implemented and connected interrupt lines for each interrupt controller

File: include/zephyr/drivers/intc/irq_flags.h
Comment (rruuaanng) [inline include/zephyr/drivers/intc/irq_flags.h:1]: Can you add a sample `dts` file in the comment? The binding-API does not have full context, which confuses users (including me) :(

Like this
```
 *     intc1: interrupt-controller@... {
 *             compatible = "vnd,interrupt";
 *             #interrupt-cells = <2>;
 *     };
 *
 *     intc2: interrupt-controller@... {
 *             compatible = "vnd,interrupt";
 *             #interrupt-cells = <2>;
 *     };
 *
 *     n: node {
 *             interrupts-extended = <&intc1 3 2 15>,
 *                                   <&intc2 4 3 16>;
 *     };
```

Comment (bjarki-andreasen) [reply include/zephyr/drivers/intc/irq_flags.h:1]: Added the following to the descriptions of the macros for getting IRQ flags from devicetree in `include/zephyr/drivers/intc/irq_flags.h`:
```
 * @details The encoding of the IRQ's flags is specific to each interrupt
 * controller, tied to its devicetree compatible, defined in
 * include/zephyr/drivers/intc/irq_flags.h
```

File: doc/kernel/services/sys_irq.rst [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+on the corresponding INTL.
+
+It is the responsibility of the INTC device driver, SoC and Arch, to do any and
+all preparation required to invoke the SYS IRQ INTL handlers from a execution
+context.
+
+Any and all configuration of INTLs is performed by SYS IRQ through the INTC device
+driver API.
+
+INTD (INTerrupt generating Device)
```
Comment (rruuaanng) [inline doc/kernel/services/sys_irq.rst:34]: I think the abbreviation and the full name should be separated. Simply put, the whole process should be capitalized according to the first letter of the sentence, without deliberately showing the abbreviation.

For example
```
INTD (Interrupt generating device)
```

(Applicable to all changes.)

**Comment by rruuaanng** [comment]
@bjarki-andreasen

**Comment by ajf58** [comment]
I'm trying to write an interrupt controller for the RP2350 to support the Hazard3 (RISC-V) cores. Currently, is it fair to assume this PR isn't going to land in the next 3 months?

**Comment by bjarki-andreasen** [comment]
> I'm trying to write an interrupt controller for the RP2350 to support the Hazard3 (RISC-V) cores. Currently, is it fair to assume this PR isn't going to land in the next 3 months?

yes, I'm working full time on power management refactoring ATM :)
